apps/backend/src/services/usersServices.ts

```diff
@@ -0,0 +1,173 @@
+import bcrypt from 'bcrypt';
+import { User } from '@prisma/client';
+import { CreateUserData, UserWithoutId, UserWithoutPasswordHash } from '../types';
+import prismaClient from '../../prismaClient';
+
+const USER_FIELDS_WITHOUT_PASSWORD_HASH = {
+  id: true,
+  user_name: true,
+  first_name: true,
+  last_name: true,
+  dob: true,
+  gender_id: true,
+  bio: true,
+};
+
+// helpers
+
+/**
+ * - narrows username type to string, checks if username is valid
+ * - throws Error if username already exists
+ * - throws Error if username not valid
+ * @param username this should be a valid username
+ * @returns resolved promise with username as value
+ */
+const parseAndValidateUsername = async (username: unknown): Promise<string> => {
+  const MINIMUM_USERNAME_LENGTH = 4;
+  if (typeof username !== 'string' || username.length < MINIMUM_USERNAME_LENGTH) {
+    const error = new Error('Invalid Username');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const existingUser: User | null = await prismaClient.user.findUnique({
+    where: {
+      user_name: username,
+    },
+  });
+
+  if (existingUser) {
+    const error = new Error('User already exists');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  return Promise.resolve(username);
+};
+
+/**
+   * - narrows type of password to string, generates password hash from the password
+   * - throws error if password is missing or invalid ie less than 4 characters
+   * @param password this should be a valid password
+   * @returns Promise resolved to password hash
+   */
+const getPasswordHash = async (password: unknown): Promise<string> => {
+  const MINIMUM_PASSWORD_LENGTH = 4;
+  if (typeof password !== 'string' || password.length < MINIMUM_PASSWORD_LENGTH) {
+    const error = new Error('Invalid Password');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const SALT_ROUNDS = 10;
+  const passwordHash = await bcrypt.hash(password, SALT_ROUNDS);
+  return Promise.resolve(passwordHash);
+};
+
+// user services functions
+
+/**
+ * - performs type narrowing for data from req.body to POST /user
+ * - if username, password present and of type string, then returns an body containing input data
+ * - the mandatory fields can be changed in future
+ * - else throws Error
+ * @param body request.body should contain username and password
+ * @returns Promise resolved to user input to POST /user
+ */
+const parseCreateUserData = async (body: unknown): Promise<CreateUserData> => {
+  if (!body || typeof body !== 'object') {
+    const error = new Error('unable to parse data');
+    error.name = 'InvalidInputError';
+    throw error;
+  }
+
+  if ('username' in body && 'password' in body
+      && typeof body.username === 'string'
+      && typeof body.password === 'string') {
+    const createUserData = {
```

Comment:
We could set a type to `CreateUserData` here.

Reply:
done